pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1374,3 +1647,95 @@ func makeDirectoryInDatastore(c *govmomi.Client, dc *object.Datacenter, path str
 
 	return err
 }
+
+// Gets the folder path from the VM Disk path.
+func getFolderFromVMDiskPath(vmDiskPath string) string {
+	i := strings.Index(vmDiskPath, "/")
+	folderPath := ""
+	if i > -1 {
+		folderPath = vmDiskPath[:i+1]
+	}
+	return folderPath
+}
+
+// Delete the VM.
+// An attempt is made to delete the VM for DeleteVMAttempts times before it backs off.
+func deleteVM(ctx context.Context, vm *object.VirtualMachine) error {
+	var destroyTask *object.Task
+	var err error
+	for i := 0; i < DeleteVMAttempts; i++ {
```

Comment:
I too don't like retry loop. Scheduling background job is great. And if its not an option then we should try to delete  on next provisioning call?